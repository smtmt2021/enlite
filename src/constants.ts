/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
const isTesting = process.env.NODE_ENV === 'test';
export const TIMEOUT_GET = (isTesting ? 1 : 10) * 1000;
