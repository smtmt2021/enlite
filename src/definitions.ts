/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export const PORT_ENLITE = 3610;
