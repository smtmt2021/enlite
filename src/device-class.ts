/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export enum CLASS {
  NODE_PROFILE = 0x0ef0,
  CONTROLLER = 0x05ff,
  LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER = 0x0288
}

export type ClassType = CLASS.NODE_PROFILE | CLASS.CONTROLLER | CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER;
