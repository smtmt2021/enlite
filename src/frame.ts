/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { FrameParams } from './interfaces';
import { Property } from './property';
import { EPC } from './epc';
import { ESV } from './esv';
import { num2hex, dumpFrameParams, getResponseService } from './lib';

const EHD = 0x1081;

export class Frame {
  static create(params: FrameParams): Frame {
    if (!params.tid || !params.seoj || !params.deoj || !params.esv) {
      throw new Error(`Frame: params are not enough ${dumpFrameParams(params)}`);
    }
    let data = num2hex(EHD, 2);
    data += num2hex(params.tid, 2);
    data += num2hex(params.seoj, 3);
    data += num2hex(params.deoj, 3);
    data += num2hex(params.esv, 1);
    data += num2hex(params.properties.length, 1); // as OPC
    for (const property of params.properties) {
      data += num2hex(property.epc, 1) + num2hex(property.pdc, 1);
      if (property.edt) {
        for (const dt of property.edt.values()) {
          data += num2hex(dt, 1);
        }
      }
    }
    return new Frame(Buffer.from(data, 'hex'));
  }

  static confirmed(frame: Frame): FrameParams {
    const properties: Property[] = [];
    for (const property of frame.properties) {
      properties.push(Property.confirmed(property));
    }
    return {
      tid: frame.tid,
      seoj: frame.deoj,
      deoj: frame.seoj,
      esv: getResponseService(frame.esv),
      properties
    };
  }

  constructor(public readonly buffer: Buffer) {}

  get ehd(): number {
    return this.buffer.readUInt16BE(0);
  }

  set tid(tid: number) {
    this.buffer.writeUInt16BE(tid, 2);
  }

  get tid(): number {
    return this.buffer.readUInt16BE(2);
  }

  get seoj(): number {
    return this.buffer.readUIntBE(4, 3);
  }

  get deoj(): number {
    return this.buffer.readUIntBE(7, 3);
  }

  get esv(): ESV {
    return this.buffer.readUInt8(10) as ESV;
  }

  get opc(): number {
    return this.buffer.readUInt8(11);
  }

  get properties(): Property[] {
    const properties: Property[] = [];
    const buffer = this.buffer.subarray(12);
    let offset = 0;
    for (let i = 0; i < this.opc; ++i) {
      const epc = buffer[offset++] as EPC;
      const pdc = buffer[offset++];
      const edt = pdc ? buffer.subarray(offset, offset + pdc) : undefined;
      properties.push(new Property(epc, edt));
      offset += pdc;
    }
    return properties;
  }

  toString(): string {
    let result = 'ehd:' + num2hex(this.ehd, 2)
      + ' tid:' + num2hex(this.tid, 2)
      + ' seoj:' + num2hex(this.seoj, 3)
      + ' deoj:' + num2hex(this.deoj, 3)
      + ' esv:' + num2hex(this.esv, 1)
      + ' opc:' + num2hex(this.opc, 1)
      + ' ';
    if (this.opc) {
      for (const property of this.properties) {
        result += property.toString();
      }
    }
    return result;
  }
}
