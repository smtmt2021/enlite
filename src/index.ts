/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { Enlite } from './enlite';
import { EnliteImpl } from './enlite-impl';
export { Enlite };
export { SendToRemote as Controller } from './interfaces';
export { Device } from './device';
export { CLASS } from './device-class';
export { SmartElectricEnergyMeter } from './smart-electric-energy-meter';

export function EnliteGetInstance(): Enlite {
  return EnliteImpl.getInstance();
}
