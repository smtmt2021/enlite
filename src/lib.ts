/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { FrameParams, TimeValue } from './interfaces';
import { ESV } from './esv';

/**
 *  Convert number to hex-decimal string.
 *  @param n     to be converted.
 *  @param bytes number of bytes, e.g. 2 bytes means 4 digits.
 */
export function num2hex(n: number, bytes = 2): string {
  if (n > Number.MAX_SAFE_INTEGER) {
    throw new Error(`the specified number is too large, ${n}`);
  }
  if (bytes < 1 || bytes > 8) {
    throw new TypeError(`invalid size of bytes, $${bytes}`);
  }
  return n
    .toString(16)
    .toUpperCase()
    .padStart(bytes * 2, '0');
}

export function dumpFrameParams(params: FrameParams): string {
  let result = '';
  if (params.tid) {
    result += `tid:${num2hex(params.tid, 2)} `;
  }
  if (params.seoj) {
    result += `seoj:${num2hex(params.seoj, 3)} `;
  }
  if (params.deoj) {
    result += `deoj:${num2hex(params.deoj, 3)} `;
  }
  for (const property of params.properties) {
    result += property.toString();
  }
  return result;
}

export function getResponseService(esv: ESV): ESV {
  switch (esv) {
    case ESV.GET:
      return ESV.GET_RES;
    case ESV.INFC:
      return ESV.INFC_RES;
  }
  return esv;
}

export function edt2DateValue(edt?: Buffer): TimeValue | undefined {
  if (!edt || edt.length !== 11) {
    return;
  }
  const year = edt.readUInt16BE(0);
  const month = edt.readUInt8(2) - 1;
  const day = edt.readUInt8(3);
  const hour = edt.readUInt8(4);
  const minuite = edt.readUInt8(5);
  const second = edt.readUInt8(6);
  const value = edt.readUInt32BE(7);
  const time = new Date(year, month, day, hour, minuite, second, 0);
  return { time, value };
}

export function dateValue2Buffer(time: Date, value: number): Buffer {
  const buffer = Buffer.alloc(11);
  buffer.writeUInt16BE(time.getFullYear());
  buffer.writeUInt8(time.getMonth() + 1, 2);
  buffer.writeUInt8(time.getDate(), 3);
  buffer.writeUInt8(time.getHours(), 4);
  buffer.writeUInt8(time.getMinutes(), 5);
  buffer.writeUInt8(time.getSeconds(), 6);
  buffer.writeUInt32BE(value, 7);
  return buffer;
}
