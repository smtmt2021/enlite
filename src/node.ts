/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { Frame } from './frame';
import { ESV } from './esv';
import { CLASS } from './device-class';
import DEBUG from 'debug';
const debug = DEBUG('enlite/node');

export abstract class Node {
  private active = false;
  protected tid = 0;
  get id(): number {
    return this.deviceClass * 0x100 + this.instanceCode;
  }

  get isOpen(): boolean {
    return this.active;
  }

  constructor(
    protected readonly deviceClass: CLASS,
    protected readonly instanceCode = 1
  ) {
    if (instanceCode < 1 || instanceCode > 0xff) {
      throw new Error('enlite: invalid instance code');
    }
    debug('node.created', this.constructor.name);
  }

  open(): Promise<boolean> {
    debug('node.open', this.constructor.name);
    if (this.active) {
      console.info(`enlite: ${this.constructor.name} is already opened`);
      return Promise.resolve(false);
    }
    this.active = true;
    return Promise.resolve(true);
  }

  close(): Promise<boolean> {
    debug('node.close', this.constructor.name);
    if (!this.active) {
      console.info(`enlite: ${this.constructor.name} is not opened`);
      return Promise.resolve(false);
    }
    this.active = false;
    return Promise.resolve(true);
  }

  protected getNextTid(): number {
    this.tid += 1;
    this.tid &= 0xffff;
    return this.tid;
  }

  protected onGetRes(address: string, frame: Frame): Promise<void> {
    debug('node.onGetRes', address, frame.toString());
    return Promise.resolve();
  }

  protected onInf(address: string, frame: Frame): Promise<void> {
    debug('node.onInf', address, frame.toString());
    return Promise.resolve();
  }

  protected onInfC(address: string, frame: Frame): Promise<void> {
    debug('node.onInfC', address, frame.toString());
    return Promise.resolve();
  }

  onData(address: string, frame: Frame): Promise<void> {
    if (!this.isOpen) {
      return Promise.resolve();
    }
    switch (frame.esv) {
      case ESV.GET_RES:
        return this.onGetRes(address, frame);
      case ESV.INF:
        return this.onInf(address, frame);
      case ESV.INFC:
        return this.onInfC(address, frame);
    }
    debug(`node.onData: unknown esv, ${frame.toString()}`);
    return Promise.resolve();
  }
}
