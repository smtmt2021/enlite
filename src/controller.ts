/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { SendToRemote, FrameParams } from './interfaces';
import { Node } from './node';
import { Frame } from './frame';
import { Device } from './device';
import { CLASS } from './device-class';
import DEBUG from 'debug';
const debug = DEBUG('enlite/controller');

export class Controller extends Node implements SendToRemote {
  protected deviceList: Device[] = [];

  constructor(private nodeProfile: SendToRemote) {
    super(CLASS.CONTROLLER);
    debug('controller created');
  }

  async close(): Promise<boolean> {
    debug('controller.close');
    for (const device of this.deviceList) {
      await device.close();
    }
    this.deviceList = [];
    return super.close();
  }

  onData(address: string, frame: Frame): Promise<void> {
    debug('controller.onData', address, frame.seoj.toString(16));
    const device = this.lookup(address, frame.seoj);
    return device ? device.onData(address, frame) : Promise.resolve();
  }

  send(params: FrameParams, address: string): Promise<void> {
    debug('controller.send');
    if (!params.deoj || !params.esv) {
      throw new Error(`Controller.send: params are not enough, ${JSON.stringify(params)}`);
    }
    params.seoj = this.id;
    return this.nodeProfile.send(params, address);
  }

  addDevice(device: Device): Promise<boolean> {
    debug('controller.addDevice');
    if (this.lookup(device.address, device.id)) {
      console.info('enlite: the device is already added', device.id.toString(16), device.address);
      return Promise.resolve(false);
    }
    this.deviceList.push(device);
    return Promise.resolve(true);
  }

  private lookup(address: string, seoj: number): Device | undefined {
    debug('controller.lookup', address, seoj.toString(16));
    return this.deviceList.find(
      device => device.address === address && device.id === seoj
    );
  }
}
