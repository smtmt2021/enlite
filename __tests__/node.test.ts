/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { Node } from '../src/node';
import { Frame } from '../src/frame';
import { CLASS } from '../src/device-class';

class NodeTest extends Node {
  getNextTid(): number {
    return super.getNextTid();
  }

  onGetRes = jest.fn((address: string, frame: Frame) => super.onGetRes(address, frame));
  onInf = jest.fn((address: string, frame: Frame) => super.onInf(address, frame));
  onInfC = jest.fn((address: string, frame: Frame) => super.onInfC(address, frame));
}

test('creatting a node', () => {
  const node1 = new NodeTest(CLASS.NODE_PROFILE);
  expect(node1.id).toEqual(CLASS.NODE_PROFILE * 0x100 + 1);

  const node2 = new NodeTest(CLASS.CONTROLLER, 1);
  expect(node2.id).toEqual(CLASS.CONTROLLER * 0x100 + 1);

  const node3 = new NodeTest(CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER, 2);
  expect(node3.id).toEqual(CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER * 0x100 + 2);

  expect(() => new NodeTest(CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER, 0x100)).toThrowErrorMatchingSnapshot();
  expect(() => new NodeTest(CLASS.NODE_PROFILE, 0)).toThrowErrorMatchingSnapshot();
});

test('open and close', async () => {
  const node = new NodeTest(CLASS.NODE_PROFILE);
  await expect(node.open()).resolves.toBeTruthy();
  await expect(node.close()).resolves.toBeTruthy();
});

test('tid is incremented one by one', () => {
  const node = new NodeTest(CLASS.NODE_PROFILE);
  expect(node.getNextTid()).toEqual(1);
  expect(node.getNextTid()).toEqual(2);
  for (let i = 0; i < 0xfffc; ++i) {
    node.getNextTid();
  }
  expect(node.getNextTid()).toEqual(0x0ffff);
  expect(node.getNextTid()).toEqual(0);
});

test('onData to onInf', async () => {
  const node = new NodeTest(CLASS.NODE_PROFILE);
  const frame = new Frame(Buffer.from('1081006B0EF0010EF0017301D50401028801', 'hex'));
  const address = 'fe80:0000:0000:0000:0000:0000:0000:0001';
  await expect(node.open()).resolves.toBeTruthy();
  await expect(node.onData(address, frame)).resolves.toBeUndefined();
  expect(node.onGetRes.mock.calls.length).toBe(0);
  expect(node.onInf.mock.calls.length).toBe(1);
  expect(node.onInfC.mock.calls.length).toBe(0);
  expect(node.onInf.mock.calls[0][0]).toBe(address);
  expect(node.onInf.mock.calls[0][1]).toBe(frame);
  await expect(node.close()).resolves.toBeTruthy();
});

test('onData to onGetRes', async () => {
  const node = new NodeTest(CLASS.NODE_PROFILE);
  const frame = new Frame(Buffer.from('1081000205ff010288017201e700', 'hex'));
  const address = 'fe80:0000:0000:0000:0000:0000:0000:0002';
  await expect(node.open()).resolves.toBeTruthy();
  await expect(node.onData(address, frame)).resolves.toBeUndefined();
  expect(node.onGetRes.mock.calls.length).toBe(1);
  expect(node.onInf.mock.calls.length).toBe(0);
  expect(node.onInfC.mock.calls.length).toBe(0);
  expect(node.onGetRes.mock.calls[0][0]).toBe(address);
  expect(node.onGetRes.mock.calls[0][1]).toBe(frame);
  await expect(node.close()).resolves.toBeTruthy();
});

test('onData to onInfC', async () => {
  const node = new NodeTest(CLASS.NODE_PROFILE);
  const frame = new Frame(Buffer.from('1081029F02880105FF017402EA0B07E501121200000000FBD2EB0B07E5011212000000000009', 'hex'));
  const address = 'fe80:0000:0000:0000:0000:0000:0000:0003';
  await expect(node.open()).resolves.toBeTruthy();
  await expect(node.onData(address, frame)).resolves.toBeUndefined();
  expect(node.onGetRes.mock.calls.length).toBe(0);
  expect(node.onInf.mock.calls.length).toBe(0);
  expect(node.onInfC.mock.calls.length).toBe(1);
  expect(node.onInfC.mock.calls[0][0]).toBe(address);
  expect(node.onInfC.mock.calls[0][1]).toBe(frame);
  await expect(node.close()).resolves.toBeTruthy();
});

test('onData to Get?', async () => {
  const node = new NodeTest(CLASS.NODE_PROFILE);
  const frame = new Frame(Buffer.from('1081000205ff010288016201e700', 'hex'));
  const address = 'fe80:0000:0000:0000:0000:0000:0000:0004';
  await expect(node.open()).resolves.toBeTruthy();
  await expect(node.onData(address, frame)).resolves.toBeUndefined();
  expect(node.onGetRes.mock.calls.length).toBe(0);
  expect(node.onInf.mock.calls.length).toBe(0);
  expect(node.onInfC.mock.calls.length).toBe(0);
  await expect(node.close()).resolves.toBeTruthy();
});
