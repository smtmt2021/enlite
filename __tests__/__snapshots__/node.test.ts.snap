// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creatting a node 1`] = `"enlite: invalid instance code"`;

exports[`creatting a node 2`] = `"enlite: invalid instance code"`;
